test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+            out_subarr_idxs[1],
+            out_subarr_idxs[2],
+            out_subarr_idxs[3]);
+
+        // The output array that will be given to the client is the subarray,
+        // not the full array
+        af_retain_array(&out_arr_c, subarr.get());
+    }
+
+    af_array getOutputArrayC() {
+        return out_arr_c;
+    }
+
+    void setOutputArray(af::array array) {
+        // When reassigning the output array, release the old retained af_array
+        // then retain the new af_array of the new C++ array
+        af_release_array(out_arr_c);
+        out_arr = array;
+        af_retain_array(&out_arr_c, out_arr.get());
+    }
+
+    // This is used in the case of NULL_ARRAY, when taking the output array
+    // generated by the af_* function. Doing this gives TestOutputArrayInfo the
+    // responsibility to release that output array
+    void setOutputArrayC(af_array array) {
+        out_arr_c = array;
+    }
+
+    af::index getSubArrayIdx(dim_t dim) {
+        return out_subarr_idxs[dim];
+    }
+};
+
+// Generates a null af_array. testWriteToOutputArray expects that the array that
+// it receives after the af_* function is a valid, allocated af_array
+void genNullArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    metadata.initArrays();
+}
+
+// Generates a random array. testWriteToOutputArray expects that it will receive
+// the same af_array that this generates after the af_* function is called
+void genRegularArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    metadata.initArrays(dims, ty);
+}
+
+// Generates a large, random array, and extracts a subarray for the af_* function
+// to use. testWriteToOutputArray expects that the large array that it receives is
+// equal to the same large array with the gold array injected on the same subarray location
+void genSubArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    const dim_t pad_size = 2;
+
+    // The large array is padded on both sides of each dimension
+    // Padding is only applied if the dimension is used, i.e. if dims[i] > 1
+    af::dim4 full_arr_dims(dims[0] > 1 ? dims[0] + 2*pad_size : dims[0],
+                           dims[1] > 1 ? dims[1] + 2*pad_size : dims[1],
+                           dims[2] > 1 ? dims[2] + 2*pad_size : dims[2],
+                           dims[3] > 1 ? dims[3] + 2*pad_size : dims[3]);
+
+    // Calculate index of sub-array. These will be used also by
+    // testWriteToOutputArray so that the gold sub array will be placed in the
+    // same location. Currently, this location is the center of the large array
+    af::index subarr_idxs[4];
+    for (int i = 0; i < 4; ++i) {
+        subarr_idxs[i] =
+            dims[i] > 1 ? af::seq(pad_size, pad_size + dims[i] - 1) : af::span;
+    }
+
+    metadata.initArrays(full_arr_dims, ty, subarr_idxs);
```

Comment:
Will become `metadata.init(full_arr_dims, ty, subarr_idxs)`

Reply:
Same concern as above (`genRegularArray`). In addition, however, I think we also need to store a copy of the original random full array that we generated, because in this case, we test if the output of `af_approx1` modifies the subarray and only the subarray - the rest of the full array must be untouched. Thus if we make a copy of the original random array that we generated, and we compare the original and the copy after (1) `af_approx1` places its output on the subarray of the original and after (2) we inject the reference values on the subarray of the copy - the original and the copy full arrays must have the same values element-wise.